In [251]:
import pandas as pd

In [252]:
white = pd.read_csv('winequality-white.csv')
red = pd.read_csv('winequality-red.csv')
white
red

fixed acidity  volatile acidity  citric acid  residual sugar  chlorides  \
0               7.4             0.700         0.00             1.9      0.076   
1               7.8             0.880         0.00             2.6      0.098   
2               7.8             0.760         0.04             2.3      0.092   
3              11.2             0.280         0.56             1.9      0.075   
4               7.4             0.700         0.00             1.9      0.076   
...             ...               ...          ...             ...        ...   
1594            6.2             0.600         0.08             2.0      0.090   
1595            5.9             0.550         0.10             2.2      0.062   
1596            6.3             0.510         0.13             2.3      0.076   
1597            5.9             0.645         0.12             2.0      0.075   
1598            6.0             0.310         0.47             3.6      0.067   

      free sulfur dioxide  total sulfur dioxide  density    pH  sulphates  \
0                    11.0                  34.0  0.99780  3.51       0.56   
1                    25.0                  67.0  0.99680  3.20       0.68   
2                    15.0                  54.0  0.99700  3.26       0.65   
3                    17.0                  60.0  0.99800  3.16       0.58   
4                    11.0                  34.0  0.99780  3.51       0.56   
...                   ...                   ...      ...   ...        ...   
1594                 32.0                  44.0  0.99490  3.45       0.58   
1595                 39.0                  51.0  0.99512  3.52       0.76   
1596                 29.0                  40.0  0.99574  3.42       0.75   
1597                 32.0                  44.0  0.99547  3.57       0.71   
1598                 18.0                  42.0  0.99549  3.39       0.66   

      alcohol  quality  
0         9.4        5  
1         9.8        5  
2         9.8        5  
3         9.8        6  
4         9.4        5  
...       ...      ...  
1594     10.5        5  
1595     11.2        6  
1596     11.0        6  
1597     10.2        5  
1598     11.0        6  

[1599 rows x 12 columns]

In [253]:
white['color'] = 0
white

fixed acidity  volatile acidity  citric acid  residual sugar  chlorides  \
0               7.0              0.27         0.36            20.7      0.045   
1               6.3              0.30         0.34             1.6      0.049   
2               8.1              0.28         0.40             6.9      0.050   
3               7.2              0.23         0.32             8.5      0.058   
4               7.2              0.23         0.32             8.5      0.058   
...             ...               ...          ...             ...        ...   
4893            6.2              0.21         0.29             1.6      0.039   
4894            6.6              0.32         0.36             8.0      0.047   
4895            6.5              0.24         0.19             1.2      0.041   
4896            5.5              0.29         0.30             1.1      0.022   
4897            6.0              0.21         0.38             0.8      0.020   

      free sulfur dioxide  total sulfur dioxide  density    pH  sulphates  \
0                    45.0                 170.0  1.00100  3.00       0.45   
1                    14.0                 132.0  0.99400  3.30       0.49   
2                    30.0                  97.0  0.99510  3.26       0.44   
3                    47.0                 186.0  0.99560  3.19       0.40   
4                    47.0                 186.0  0.99560  3.19       0.40   
...                   ...                   ...      ...   ...        ...   
4893                 24.0                  92.0  0.99114  3.27       0.50   
4894                 57.0                 168.0  0.99490  3.15       0.46   
4895                 30.0                 111.0  0.99254  2.99       0.46   
4896                 20.0                 110.0  0.98869  3.34       0.38   
4897                 22.0                  98.0  0.98941  3.26       0.32   

      alcohol  quality  color  
0         8.8        6      0  
1         9.5        6      0  
2        10.1        6      0  
3         9.9        6      0  
4         9.9        6      0  
...       ...      ...    ...  
4893     11.2        6      0  
4894      9.6        5      0  
4895      9.4        6      0  
4896     12.8        7      0  
4897     11.8        6      0  

[4898 rows x 13 columns]

In [254]:
red['color'] = 1
red

fixed acidity  volatile acidity  citric acid  residual sugar  chlorides  \
0               7.4             0.700         0.00             1.9      0.076   
1               7.8             0.880         0.00             2.6      0.098   
2               7.8             0.760         0.04             2.3      0.092   
3              11.2             0.280         0.56             1.9      0.075   
4               7.4             0.700         0.00             1.9      0.076   
...             ...               ...          ...             ...        ...   
1594            6.2             0.600         0.08             2.0      0.090   
1595            5.9             0.550         0.10             2.2      0.062   
1596            6.3             0.510         0.13             2.3      0.076   
1597            5.9             0.645         0.12             2.0      0.075   
1598            6.0             0.310         0.47             3.6      0.067   

      free sulfur dioxide  total sulfur dioxide  density    pH  sulphates  \
0                    11.0                  34.0  0.99780  3.51       0.56   
1                    25.0                  67.0  0.99680  3.20       0.68   
2                    15.0                  54.0  0.99700  3.26       0.65   
3                    17.0                  60.0  0.99800  3.16       0.58   
4                    11.0                  34.0  0.99780  3.51       0.56   
...                   ...                   ...      ...   ...        ...   
1594                 32.0                  44.0  0.99490  3.45       0.58   
1595                 39.0                  51.0  0.99512  3.52       0.76   
1596                 29.0                  40.0  0.99574  3.42       0.75   
1597                 32.0                  44.0  0.99547  3.57       0.71   
1598                 18.0                  42.0  0.99549  3.39       0.66   

      alcohol  quality  color  
0         9.4        5      1  
1         9.8        5      1  
2         9.8        5      1  
3         9.8        6      1  
4         9.4        5      1  
...       ...      ...    ...  
1594     10.5        5      1  
1595     11.2        6      1  
1596     11.0        6      1  
1597     10.2        5      1  
1598     11.0        6      1  

[1599 rows x 13 columns]

In [255]:
wine = pd.concat([white, red], ignore_index=True)
wine

fixed acidity  volatile acidity  citric acid  residual sugar  chlorides  \
0               7.0             0.270         0.36            20.7      0.045   
1               6.3             0.300         0.34             1.6      0.049   
2               8.1             0.280         0.40             6.9      0.050   
3               7.2             0.230         0.32             8.5      0.058   
4               7.2             0.230         0.32             8.5      0.058   
...             ...               ...          ...             ...        ...   
6492            6.2             0.600         0.08             2.0      0.090   
6493            5.9             0.550         0.10             2.2      0.062   
6494            6.3             0.510         0.13             2.3      0.076   
6495            5.9             0.645         0.12             2.0      0.075   
6496            6.0             0.310         0.47             3.6      0.067   

      free sulfur dioxide  total sulfur dioxide  density    pH  sulphates  \
0                    45.0                 170.0  1.00100  3.00       0.45   
1                    14.0                 132.0  0.99400  3.30       0.49   
2                    30.0                  97.0  0.99510  3.26       0.44   
3                    47.0                 186.0  0.99560  3.19       0.40   
4                    47.0                 186.0  0.99560  3.19       0.40   
...                   ...                   ...      ...   ...        ...   
6492                 32.0                  44.0  0.99490  3.45       0.58   
6493                 39.0                  51.0  0.99512  3.52       0.76   
6494                 29.0                  40.0  0.99574  3.42       0.75   
6495                 32.0                  44.0  0.99547  3.57       0.71   
6496                 18.0                  42.0  0.99549  3.39       0.66   

      alcohol  quality  color  
0         8.8        6      0  
1         9.5        6      0  
2        10.1        6      0  
3         9.9        6      0  
4         9.9        6      0  
...       ...      ...    ...  
6492     10.5        5      1  
6493     11.2        6      1  
6494     11.0        6      1  
6495     10.2        5      1  
6496     11.0        6      1  

[6497 rows x 13 columns]

In [256]:
red.isna().sum()

fixed acidity           0
volatile acidity        0
citric acid             0
residual sugar          0
chlorides               0
free sulfur dioxide     0
total sulfur dioxide    0
density                 0
pH                      0
sulphates               0
alcohol                 0
quality                 0
color                   0
dtype: int64

In [257]:
red_density = red['density']
white_density = white['density']

In [258]:
density = pd.concat([red_density, white_density], axis = 1)
density.columns = ['red', 'white']
density

red    white
0     0.9978  1.00100
1     0.9968  0.99400
2     0.9970  0.99510
3     0.9980  0.99560
4     0.9978  0.99560
...      ...      ...
4893     NaN  0.99114
4894     NaN  0.99490
4895     NaN  0.99254
4896     NaN  0.98869
4897     NaN  0.98941

[4898 rows x 2 columns]

In [259]:
red_density.describe()

count    1599.000000
mean        0.996747
std         0.001887
min         0.990070
25%         0.995600
50%         0.996750
75%         0.997835
max         1.003690
Name: density, dtype: float64

In [260]:
white_density.describe()

count    4898.000000
mean        0.994027
std         0.002991
min         0.987110
25%         0.991723
50%         0.993740
75%         0.996100
max         1.038980
Name: density, dtype: float64

## 가설1. 레드와인의 무게감이 화이트와인보다 클 것이다
> 더 풍부하기 때문

## 등분산 검정

>귀무가설 : 두 집단의 분산이 같다  
>대립가설 : 두 집단의 분산이 다르다

In [261]:
from scipy.stats import bartlett
import numpy as np

bartlett(red_density, white_density)

BartlettResult(statistic=430.5766981952509, pvalue=1.2172223684463466e-95)

#### Welch's T-test (등분산 가정을 만족하지 않더라도 사용 가능)

>귀무가설 : 레드와인과 화이트와인의 무게감 차이가 없다  
>대립가설 : 레드와인과 화이트와인의 무게감 차이가 있다

In [262]:
import scipy.stats as stats

In [263]:
t_stat, p_value = stats.ttest_ind(red_density, white_density, equal_var = False)
t_stat, p_value

(42.70851107543416, 0.0)

상관계수가 가장 높은 변수가 (무게감, 알콜)  
두번째로 높은 변수가 (무게감, 잔여 설탕)  
따라서 다중공선성 문제를 제거하기 위해 무게감 컬럼을 제거하기로 결정

## 독립변수 표준화

In [264]:
X = wine.iloc[:, :-1] # 독립변수
y = wine['color'] # 종속변수

In [265]:
from sklearn.preprocessing import StandardScaler

scaler = StandardScaler()

X_scaled = scaler.fit_transform(X)

In [266]:
X_scaled

array([[-0.16608919, -0.42318303,  0.28468605, ..., -0.54617826,
        -1.41855821,  0.20799905],
       [-0.70607349, -0.24094936,  0.14704613, ..., -0.27735097,
        -0.83161516,  0.20799905],
       [ 0.68245757, -0.36243847,  0.55996589, ..., -0.61338508,
        -0.32852111,  0.20799905],
       ...,
       [-0.70607349,  1.03468634, -1.29817304, ...,  1.47002637,
         0.42611996,  0.20799905],
       [-1.01463595,  1.85473786, -1.366993  , ...,  1.20119908,
        -0.2446721 , -0.93722961],
       [-0.93749534, -0.1802048 ,  1.04170561, ...,  0.86516498,
         0.42611996,  0.20799905]])

In [267]:
import statsmodels.api as sm

X = wine.iloc[:, :-1] # 독립변수
y = wine['color'] # 종속변수

# 상수항 추가 (절편)
X2 = sm.add_constant(X)

# 회귀 분석 모델 생성 및 적합
model = sm.OLS(y, X2)
results = model.fit()

# 회귀 분석 결과 출력
print(results.summary())

                            OLS Regression Results                            
Dep. Variable:                  color   R-squared:                       0.862
Model:                            OLS   Adj. R-squared:                  0.862
Method:                 Least Squares   F-statistic:                     3388.
Date:                Thu, 22 Aug 2024   Prob (F-statistic):               0.00
Time:                        20:16:38   Log-Likelihood:                 2697.2
No. Observations:                6497   AIC:                            -5368.
Df Residuals:                    6484   BIC:                            -5280.
Df Model:                          12                                         
Covariance Type:            nonrobust                                         
                           coef    std err          t      P>|t|      [0.025      0.975]
----------------------------------------------------------------------------------------
const                 -135.5378 

In [268]:
import statsmodels.api as sm


# 상수항 추가 (절편)
X_scaled = sm.add_constant(X_scaled)

# 회귀 분석 모델 생성 및 적합
model = sm.OLS(y, X_scaled)
results = model.fit()

# 회귀 분석 결과 출력
print(results.summary())

                            OLS Regression Results                            
Dep. Variable:                  color   R-squared:                       0.862
Model:                            OLS   Adj. R-squared:                  0.862
Method:                 Least Squares   F-statistic:                     3388.
Date:                Thu, 22 Aug 2024   Prob (F-statistic):               0.00
Time:                        20:16:38   Log-Likelihood:                 2697.2
No. Observations:                6497   AIC:                            -5368.
Df Residuals:                    6484   BIC:                            -5280.
Df Model:                          12                                         
Covariance Type:            nonrobust                                         
                 coef    std err          t      P>|t|      [0.025      0.975]
------------------------------------------------------------------------------
const          0.2461      0.002    124.049      0.0

In [269]:
from sklearn.linear_model import LogisticRegression

# 로지스틱 회귀 모델 생성 (다중 클래스)
model = LogisticRegression(multi_class='multinomial', solver='lbfgs', max_iter=1000)

# 모델 학습
model.fit(X_scaled, y)

/opt/anaconda3/envs/PY_39/lib/python3.9/site-packages/sklearn/linear_model/_logistic.py:1237: FutureWarning: 'multi_class' was deprecated in version 1.5 and will be removed in 1.7. From then on, binary problems will be fit as proper binary  logistic regression models (as if multi_class='ovr' were set). Leave it to its default value to avoid this warning.
  warnings.warn(


LogisticRegression(max_iter=1000, multi_class='multinomial')

In [270]:
# 회귀 계수와 절편 출력
coefficients = model.coef_
intercept = model.intercept_

print("Coefficients:\n", coefficients)
print("Intercept:\n", intercept)

Coefficients:
 [[-0.81870787 -0.0667002   0.58906111 -0.16987952 -1.84111459  0.3986594
   0.44401473 -1.41245619  2.27055973 -0.00932989  0.25242216  0.80574136
   0.17419265]]
Intercept:
 [-1.09809318]


In [271]:
# 출력 예시 (임의로 입력된 값)
coefficients = model.coef_  # 예: array([[1.5, -0.2, 0.3], [-0.5, 0.3, -0.1]])
intercept = model.intercept_  # 예: array([0.2, -0.1, 0.5])

print("Coefficients:\n", coefficients)
print("Intercept:\n", intercept)

Coefficients:
 [[-0.81870787 -0.0667002   0.58906111 -0.16987952 -1.84111459  0.3986594
   0.44401473 -1.41245619  2.27055973 -0.00932989  0.25242216  0.80574136
   0.17419265]]
Intercept:
 [-1.09809318]


In [272]:
X

fixed acidity  volatile acidity  citric acid  residual sugar  chlorides  \
0               7.0             0.270         0.36            20.7      0.045   
1               6.3             0.300         0.34             1.6      0.049   
2               8.1             0.280         0.40             6.9      0.050   
3               7.2             0.230         0.32             8.5      0.058   
4               7.2             0.230         0.32             8.5      0.058   
...             ...               ...          ...             ...        ...   
6492            6.2             0.600         0.08             2.0      0.090   
6493            5.9             0.550         0.10             2.2      0.062   
6494            6.3             0.510         0.13             2.3      0.076   
6495            5.9             0.645         0.12             2.0      0.075   
6496            6.0             0.310         0.47             3.6      0.067   

      free sulfur dioxide  total sulfur dioxide  density    pH  sulphates  \
0                    45.0                 170.0  1.00100  3.00       0.45   
1                    14.0                 132.0  0.99400  3.30       0.49   
2                    30.0                  97.0  0.99510  3.26       0.44   
3                    47.0                 186.0  0.99560  3.19       0.40   
4                    47.0                 186.0  0.99560  3.19       0.40   
...                   ...                   ...      ...   ...        ...   
6492                 32.0                  44.0  0.99490  3.45       0.58   
6493                 39.0                  51.0  0.99512  3.52       0.76   
6494                 29.0                  40.0  0.99574  3.42       0.75   
6495                 32.0                  44.0  0.99547  3.57       0.71   
6496                 18.0                  42.0  0.99549  3.39       0.66   

      alcohol  quality  
0         8.8        6  
1         9.5        6  
2        10.1        6  
3         9.9        6  
4         9.9        6  
...       ...      ...  
6492     10.5        5  
6493     11.2        6  
6494     11.0        6  
6495     10.2        5  
6496     11.0        6  

[6497 rows x 12 columns]

In [273]:
import pandas as pd
from sklearn.preprocessing import StandardScaler
from sklearn.linear_model import LogisticRegression
from sklearn.metrics import classification_report, confusion_matrix

# 예시 데이터 준비 (데이터프레임 X와 시리즈 y가 이미 주어졌다고 가정)
# X = pd.DataFrame({...})
# y = pd.Series([...])

# 데이터 스케일링 (수치형 변수)
scaler = StandardScaler()
X_scaled = scaler.fit_transform(X)

# 로지스틱 회귀 모델 생성
model = LogisticRegression()

# 모델 학습
model.fit(X_scaled, y)

# 예측
y_pred = model.predict(X_scaled)

# 평가
print("Confusion Matrix:")
print(confusion_matrix(y, y_pred))
print("\nClassification Report:")
print(classification_report(y, y_pred))

# 모델 계수 출력
print("Model Coefficients:")
print(model.coef_)
print("Intercept:")
print(model.intercept_)

Confusion Matrix:
[[4879   19]
 [  19 1580]]

Classification Report:
              precision    recall  f1-score   support

           0       1.00      1.00      1.00      4898
           1       0.99      0.99      0.99      1599

    accuracy                           0.99      6497
   macro avg       0.99      0.99      0.99      6497
weighted avg       0.99      0.99      0.99      6497

Model Coefficients:
[[ 0.12152678  1.20892875 -0.3364469  -3.26101619  0.83237943  0.75448585
  -2.73290463  3.9497415   0.17589864  0.57406085  1.33840432  0.31911147]]
Intercept:
[-3.81624253]


In [274]:
model.coef_

array([[ 0.12152678,  1.20892875, -0.3364469 , -3.26101619,  0.83237943,
         0.75448585, -2.73290463,  3.9497415 ,  0.17589864,  0.57406085,
         1.33840432,  0.31911147]])

In [275]:
def calc_prob(coef):
    e = np.e
    p_value = (e ** coef) / (1 + (e ** coef))
    return p_value

In [276]:
calc_prob(model.coef_)

array([[0.53034436, 0.77010935, 0.41667283, 0.03693305, 0.69685781,
        0.68015536, 0.06105942, 0.98110425, 0.54386163, 0.63969967,
        0.79222741, 0.57910769]])

## 무게감이 잔여설탕이랑 알코올에 영향을 받는지

In [277]:
import pandas as pd
import statsmodels.api as sm
from sklearn.preprocessing import StandardScaler
from sklearn.model_selection import train_test_split

# 데이터프레임 생성 (예시)
# df = pd.read_csv('your_data.csv')  # 데이터 로드 예시

# 독립변수와 종속변수 선택
X = wine[['residual sugar', 'alcohol']]
y = wine['density']

# 데이터 스케일링
scaler = StandardScaler()
X_scaled = scaler.fit_transform(X)

# 상수항 추가
X_scaled = sm.add_constant(X_scaled)

# 모델 생성 및 적합
model = sm.OLS(y, X_scaled).fit()

# 결과 출력
print(model.summary())


                            OLS Regression Results                            
Dep. Variable:                density   R-squared:                       0.579
Model:                            OLS   Adj. R-squared:                  0.579
Method:                 Least Squares   F-statistic:                     4464.
Date:                Thu, 22 Aug 2024   Prob (F-statistic):               0.00
Time:                        20:16:38   Log-Likelihood:                 31336.
No. Observations:                6497   AIC:                        -6.267e+04
Df Residuals:                    6494   BIC:                        -6.265e+04
Df Model:                           2                                         
Covariance Type:            nonrobust                                         
                 coef    std err          t      P>|t|      [0.025      0.975]
------------------------------------------------------------------------------
const          0.9947   2.41e-05   4.12e+04      0.0

logit(P(Y=1))=-3.8162 + 0.12152678*고정산도 + 1.20892875*휘발성산도 - 0.3364469*구연산 - 3.26101619*잔여설탕 + 0.83237943*염화물 + 0.75448585*유리이산화황 - 2.73290463*총이산화황 + 3.9497415*밀도 + 0.17589864*pH + 0.57406085*황산염 + 1.33840432*알코올 + 0.31911147*품질

In [278]:
calc_prob()

TypeError: calc_prob() missing 1 required positional argument: 'coef'

## 따라서 고정산도가 한 단위 증가할때 화이트 와인일 확률이 53% 증가한다.